File: java-source/src/main/java/com/example/flow/ExampleFlow.java
package com.example.flow;

import co.paralleluniverse.fibers.Suspendable;
import com.google.common.collect.ImmutableList;
import kotlin.Triple;
import net.corda.core.flows.*;
import net.corda.core.identity.Party;

import java.util.List;

public class ExampleFlow {
    @InitiatingFlow
    @StartableByRPC
    public static class Initiator extends FlowLogic<Void> {
        private final Party otherParty;

        public Initiator(Party otherParty) {
            this.otherParty = otherParty;
        }

        @Suspendable
        @Override
        public Void call() throws FlowException {
            FlowSession otherPartySession = initiateFlow(otherParty);
            List<FlowSession> sessions = ImmutableList.of(otherPartySession);
           // sessions.forEach( it -> sendAMessage(it, new Triple("a", "b", "c")));
            /**
             * please find the link for more info:
             * https://stackoverflow.com/questions/50638968/in-corda-unexpected-task-state-when-running-a-flow?rq=1
             */
            for(FlowSession session : sessions){
                sendAMessage(session, new Triple("a", "b", "c"));
            }
            return null;
        }
    }

    @InitiatedBy(Initiator.class)
    public static class Acceptor extends FlowLogic<Void> {

        private final FlowSession otherPartyFlow;

        public Acceptor(FlowSession otherPartyFlow) {
            this.otherPartyFlow = otherPartyFlow;
        }

        @Suspendable
        @Override
        public Void call() throws FlowException {
            otherPartyFlow.receive(Triple.class).unwrap(it -> { return it; });
            return null;
        }
    }

    @Suspendable
    private static Void sendAMessage(FlowSession session, Triple t) {
        session.send(t);
        return null;
    }
}

File: kotlin-source/src/main/kotlin/com/example/flow/ExampleFlow.kt
package com.example.flow

import co.paralleluniverse.fibers.Suspendable
import com.example.contract.IOUContract
import com.example.contract.IOUContract.Companion.IOU_CONTRACT_ID
import com.example.flow.ExampleFlow.Acceptor
import com.example.flow.ExampleFlow.Initiator
import com.example.state.IOUState
import net.corda.core.contracts.Command
import net.corda.core.contracts.ContractState
import net.corda.core.contracts.StateAndRef
import net.corda.core.contracts.requireThat
import net.corda.core.flows.*
import net.corda.core.identity.Party
import net.corda.core.transactions.SignedTransaction
import net.corda.core.transactions.TransactionBuilder
import net.corda.core.utilities.ProgressTracker
import net.corda.core.utilities.ProgressTracker.Step
import net.corda.core.utilities.unwrap
import net.corda.finance.contracts.asset.Cash

/**
 * This flow allows two parties (the [Initiator] and the [Acceptor]) to come to an agreement about the IOU encapsulated
 * within an [IOUState].
 *
 * In our simple example, the [Acceptor] always accepts a valid IOU.
 *
 * These flows have deliberately been implemented by using only the call() method for ease of understanding. In
 * practice we would recommend splitting up the various stages of the flow into sub-routines.
 *
 * All methods called within the [FlowLogic] sub-class need to be annotated with the @Suspendable annotation.
 */
@InitiatingFlow
@StartableByRPC
class Initiator(val iouValue: Int,
                val otherParty: Party) : FlowLogic<SignedTransaction>() {
    @Suspendable
    override fun call(): SignedTransaction {
        val notary = serviceHub.networkMapCache.notaryIdentities[0]

        val goodsState: ContractState = GoodsState()
        val txCommand = Command(GoodsContract.Commands.Delivery(), listOf(ourIdentity.owningKey, otherParty.owningKey))
        val txBuilder = TransactionBuilder(notary)

        val otherPartySession = initiateFlow(otherParty)
        val tokenState = otherPartySession.receive<StateAndRef<Cash.State>>().unwrap { it }
        txBuilder.addInputState(tokenState)

        // Verify that the transaction is valid.
        txBuilder.verify(serviceHub)

        // Sign the transaction.
        val partSignedTx = serviceHub.signInitialTransaction(txBuilder)

        // Send the state to the counterparty, and receive it back with their signature.
        val otherPartyFlow = initiateFlow(otherParty)
        val fullySignedTx = subFlow(CollectSignaturesFlow(partSignedTx, setOf(otherPartyFlow), GATHERING_SIGS.childProgressTracker()))

        // Notarise and record the transaction in both parties' vaults.
        return subFlow(FinalityFlow(fullySignedTx, FINALISING_TRANSACTION.childProgressTracker()))
    }
}
